src/libexpr/symbol-table.hh

```diff
@@ -12,82 +12,92 @@ namespace nix {
 /* Symbol table used by the parser and evaluator to represent and look
    up identifiers and attributes efficiently.  SymbolTable::create()
    converts a string into a symbol.  Symbols have the property that
-   they can be compared efficiently (using a pointer equality test),
+   they can be compared efficiently (using an equality test),
    because the symbol table stores only one copy of each string. */
 
-class Symbol
+/* This class mainly exists to give us an operator<< for ostreams. We could also
+   return plain strings from SymbolTable, but then we'd have to wrap every
+   instance of a symbol that is fmt()ed, which is inconvenient and error-prone. */
+class SymbolStr
 {
-    friend class SymbolTable;
 private:
-    std::string s;
+    const std::string * s;
```

Comment:
Any reason for this change?

Reply:
Oh I see, thanks.

The constructor below feels a bit dangerous now (it seems that we don’t construct a `SymbolStr` outside of a `SymbolTable` atm, but it would be tempting to do something like `Symbol blah("somestring")` which would lead to a dangling pointer. Any objection against either making it private or making it take a `std::string *` as argument to make it more explicit that it’s expecting the string to stay alive?